### ISSUE TITLE
No way to configure ZFS event daemon

### ISSUE THREAD

**Issue by c0deaddict** [NixOS/nixpkgs#54435]
## Issue description

ZFS's event daemon (zed) can't be configured. The config file, together with some shell scripts, is extracted to from `zfsUser` to `/etc/zfs/zed.d/zed.rc` here: https://github.com/NixOS/nixpkgs/blob/1937f1ec06abd3522f8fe8eaf502e219a623cfc5/nixos/modules/tasks/filesystems/zfs.nix#L385

### Steps to reproduce

I tried to override the config file like this:

```nix
environment.etc."zfs/zed.d/zed.rc".text = "...";
```

But that fails with these errors:

```
duplicate entry zfs/zed.d/zed.rc -> /nix/store/g7pfqmkfzv08a9309cricwqak3lha57c-etc-zed.rc
mismatched duplicate entry  <-> /nix/store/g7pfqmkfzv08a9309cricwqak3lha57c-etc-zed.rc
```

That makes sense: in `zfs.nix` a whole directory is assigned to `environment.etc."zfs/zed.d".source`, overriding a file in that directory thus becomes impossible.

A possible solution would be to not copy `zed.rc`. But I can't find a way to "filter" the list of files that is copied to the `environment.etc`.

Looks like this issue is related: #25566

**Comment by c0deaddict** [comment]
Thanks for your quick response :)

Unfortunately the error remains the same:

```nix
environment.etc."zfs/zed.d/zed.rc".text = lib.mkOverride 99 "...";
```

```
...
duplicate entry zfs/zed.d/zed.rc -> /nix/store/9mkizvi55k5dj5k0iikljgbfgqkd8fnw-etc-zed.rc
mismatched duplicate entry  <-> /nix/store/9mkizvi55k5dj5k0iikljgbfgqkd8fnw-etc-zed.rc
...
```

**Comment by andrew-d** [comment]
Ah, I bet this is because one option is defined as ”.text” and the other as
“.source”. I’m on a phone and can’t test it out, but try something like:

environment.etc."zfs/zed.d/zed.rc" = lib.mkOverride 99 { text = “...”; };

Or something like that?

**Comment by c0deaddict** [comment]
That also gives the same error. I think the problem is a level deeper. In `zfs.nix`, the key "zfs/zed.d" is defined with a value of `{source = "${packages.zfsUser}/etc/zfs/zed.d/"}`. I want to override just one file, so using "zfs/zed.d/zed.rc" as key, Nix can't merge that with the already defined key "zfs/zed.d" as it's (probably?) not aware that the string key is a filesystem path that could in theory be split.

I've tried various ways to "reprhase" the line in `zfs.nix`. I thought it might be possible to override or extend the zfs derivation and put my desired config file in the source's `zed.rc` file, but haven't got that to work yet.

Another approach, along the same lines, I tried is to create a derivation that copies all files from `"${packages.zfsUser}/etc/zfs/zed.d/"` and then change the `zed.rc` file. But i'm stabbing in the dark a bit here, i don't have much experience with creating derivations and don't know what the stdenv has to offer for this kind of problem.

Anyhow, thanks for trying to help out!
